### PR TITLE
CSV file upload, parsing, and saving to DB

### PR DESCRIPTION
**Pull request by KevinSalina**
**Dependencies installed:**
Multer, sequelize, fast-csv

**Folder Structures created:**
controllers > customers
middleware > 
models > 
resources > static > assets > uploads
routes >

**Code:**
1. Configure MyQL database with Sequelize
2. Intialize and defined Sequelize model
3. Created middleware for uploading and storing CSV file
4. Created controller for uploading/importing CSV file
5. Created routes for uploading CSV file
6. Updated express server
7. Tested in postman usinf file upload

### CODE REVIEW

File: server/controllers/customers/csv.customers.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
+const model = require('../../models')
```
Comment (IgnacioOAF) [inline server/controllers/customers/csv.customers.js]: This is just a **nitpick**:
The file naming convention is related more with jQuery. I recommended following Google's style guide for naming files and be consistent with it https://google.github.io/styleguide/jsguide.html#file-name. 
for example it could be `csv_customers.js` or `csv-customers.js`

File: server/public/javascript/sanitize.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+const sanitizeEmails = (customers) => {
```
Comment (IgnacioOAF) [inline server/public/javascript/sanitize.js]: why did you decide to place this function here? Don't you think this function should be part of the customer model?

File: server/public/javascript/sanitize.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+  for (const customer of customers) {
+    customer.email = customer.email.replace(/[&\/\\#,+()$~%'":*?<>{}\s]/g, '').replace('..', '')
```
Comment (IgnacioOAF) [inline server/public/javascript/sanitize.js]: I would recommend doing this a bit different.[ What about sequelize Hooks?](https://sequelize.org/master/manual/hooks.html) This way every time we try to create a new customer, we are going to make sure the email is going to be sanitized.

File: server/routes/customer.routes.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
+const express = require('express')
+const router = express.Router()
+const csvController = require('../controllers/customers/csv.customers.js')
+const upload = require('../middleware/upload')
+
+let routes = (app) => {
+  router.post('/upload', upload.single('file'), csvController.upload);
```
Comment (IgnacioOAF) [inline server/routes/customer.routes.js]: Sweet, Middleware! Nice find!

File: server/controllers/customers/csv.customers.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+const fs = require('fs')
+const csv = require('fast-csv')
+
+const sanitizeEmails = require('../../public/javascript/sanitize')
+
+
+// Upload and save CSV file to database
+const upload = async (req, res) => {
+  try {
```
Comment (IgnacioOAF) [inline server/controllers/customers/csv.customers.js]: Aha! don't worry I used to make this mistake too. You guys don't have to refactor this, but controllers shouldn't have too much logic. Check this article [explaining why](https://medium.com/swlh/your-controllers-are-doing-too-much-this-is-how-to-simplify-them-7f7d0ea0a810)
This csv file handler could have been a utils file that its main goal is handle CSV files. A **very important** topic, single responsibility! Here is a cool [article too](https://medium.com/@learnstuff.io/single-responsibility-principle-ad3ae3e264bb)

File: server/public/javascript/sanitize.js [hunk 6]
```diff
@@ -0,0 +1,15 @@
+const sanitizeEmails = (customers) => {
+
+  for (const customer of customers) {
```
Comment (IgnacioOAF) [inline server/public/javascript/sanitize.js]: Hmmm customers is an array right? Try to use the new ES6 functions like[ **map**.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map)

**Review by IgnacioOAF — COMMENTED** [review]
This is looking great!! I left some comments on how to refactor the code. Reach out to me if you have any questions.

- Another point I wanted to mention is: I see you have created only one model `Customer`, and in it you save all the data. I mentioned in one of the comments about single responsibility principle. This explains how every element should have one purpose. Could you think how to break the data of the CSV into other models? No need to implement, but it is great to start thinking about it.

File: server/controllers/customers/csv-customers.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
+const Customer = require('../../models/Customer')
+const handleCSV = require('../../utilities/handle-csv')
+
+// Upload and save CSV file to database
+const upload = async (req, res) => {
+  try {
+    if (req.file == undefined) return res.status(400).send("Please upload a CSV file!")
```
Comment (IgnacioOAF) [inline server/controllers/customers/csv-customers.js]: nit: this if only contains an if condition, and you can change this to be an inline condition by doing this.
```
req.file && res.status(400).send("Please upload a CSV file!")
```
Otherwise, if you prefer the if condition, it is generally better to use to leave the curly brackets. It makes it easier to see what is the context of the if block.

File: server/controllers/customers/csv-customers.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+    let path = __basedir + "/resources/static/assets/uploads/" + req.file.filename
+
+    handleCSV(path, req, res)
+
+  } catch (error) {
+    console.log(error)
+    res.status(500).send("Could not upload the file: " + req.file.originalname)
```
Comment (IgnacioOAF) [inline server/controllers/customers/csv-customers.js]: we are missing the return here
```
return rest.status(500)...
```

File: server/controllers/customers/csv-customers.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+  }
+}
+
+// Get all Cusotmers
+const getCustomers = (req, res) => {
+  Customer.findAll()
+    .then(data => res.send(data))
```
Comment (IgnacioOAF) [inline server/controllers/customers/csv-customers.js:23]: this is better if you do 
```
.then(res.send)
```
what javascript does, is it calls the send function and the arguments from the callback are passed to the send method.

File: server/middleware/upload.js [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+  }
+}
+
+// Telling multer to use disk storage engine. Pass in the folder destination and filename
+const storage = multer.diskStorage({
+  destination: (req, file, cb) => {
+    cb(null, __basedir + "/resources/static/assets/uploads/")
+  },
+  filename: (req, file, cb) => {
+    console.log(file.originalname);
```
Comment (IgnacioOAF) [inline server/middleware/upload.js]: feel free to remove this console.log

File: server/controllers/customers/csv-customers.js [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+const upload = async (req, res) => {
+  try {
+    if (req.file == undefined) return res.status(400).send("Please upload a CSV file!")
+
+    let path = __basedir + "/resources/static/assets/uploads/" + req.file.filename
+
+    handleCSV(path, req, res)
+
+  } catch (error) {
+    console.log(error)
```
Comment (IgnacioOAF) [inline server/controllers/customers/csv-customers.js]: feel free to remove this console.log

File: server/models/Customer.js [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+// Model Hooks
+Customer.beforeBulkCreate((customers) => {
+  for (const customer of customers) {
+    customer.sanitizeEmail()
+  }
+})
+
+// Instance Methods
+Customer.prototype.sanitizeEmail = function () {
+  this.email = this.email.replace(/[&\/\\#,+()$~%'":*?<>{}\s]/g, '').replace('..', '')
```
Comment (IgnacioOAF) [inline server/models/Customer.js]: what would happen if the email has `hi@test....com`?
Not too oppressing to solve this edge case, but maybe you can remove the `replace('..', '')` and add `.{2,}` (try to find what this regular expression does)

File: server/models/Customer.js [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+    type: Sequelize.INTEGER
+  },
+  lastOrderDate: {
+    type: Sequelize.STRING
+  }
+});
+
+// Model Hooks
+Customer.beforeBulkCreate((customers) => {
+  for (const customer of customers) {
```
Comment (IgnacioOAF) [inline server/models/Customer.js:38]: Can we use another function to avoid mutating the original customers array? What about `.map`

File: server/utilities/handle-csv.js [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+    .on('error', (error) => {
+      throw error.message
+    })
+    .on('data', (row) => {
+      customers.push(row)
+    })
+    .on('end', () => {
+      Customer.bulkCreate(customers)
+        .then(() => {
+          return res.status(200).send({
```
Comment (IgnacioOAF) [inline server/utilities/handle-csv.js:19]: nit: if there is only one line inside a function with a return you can replace it for 
```
.then(() => (res.status(200).send({message: 'Uploaded...'})))
```
Check where else you can apply this.

**Review by IgnacioOAF — COMMENTED** [review]
Great job with this. Left some comments to make your code a bit easier to read by following the Javascript best practices.
